Docs, Trait instead of separate resource

Leinnan opened the pull request.

Leinnan committed:
Docs, some rework

extrawurst commented on README.md:
```diff
@@ -46,14 +46,23 @@ bevy_ios_safearea = { version = "=0.1.0" }
-    let safe_area_top = safe_area.top();
+use bevy::prelude::*;
+fn bevy_system(safe_area: Option<Res<bevy_ios_safearea::IosSafeArea>>) {
+    let Some(area) = safe_area else {
```
i would still keep the usage example of `top()` around which defaults to 0 on non-ios

extrawurst replied:
as like a convenience method

Leinnan replied:
Restored with a trait ;)

extrawurst commented on bevy_ios_safearea/Cargo.toml:
```diff
@@ -19,3 +18,7 @@ bevy = { version = "0.15", default-features = false, features = [
+
+
+[lints.rust]
+missing_docs = "warn"
```
please put a newline at the end

extrawurst commented on bevy_ios_safearea/src/plugin.rs:
```diff
@@ -1,35 +1,45 @@
 #[allow(dead_code)]
 #[derive(Default)]
-pub struct IosSafeAreaPlugin;
+pub struct SafeAreaPlugin;
```
its sort of convention with the other crates that the plugins all start with `Ios`

Leinnan replied:
Restored

extrawurst commented on bevy_ios_safearea/src/plugin.rs:
```diff
@@ -54,12 +64,13 @@ fn init(
-            let safe_area = IosSafeArea { top, bottom };
+            let safe_area = IosSafeArea { top, bottom, left, right, };
 
+            commands.insert_resource(safe_area);
```
as a side: how does insert-overwrite behave with change-detection?

Leinnan replied:
I think it is treated as `added` and `changed` but I don't remember now.

Leinnan committed:
Merge remote-tracking branch 'upstream/main'

# Conflicts:
#	README.md
#	bevy_ios_safearea/Cargo.lock
#	bevy_ios_safearea/Cargo.toml
#	bevy_ios_safearea/src/lib.rs

Leinnan committed:
Apply feedback

Leinnan committed:
Fix `cargo doc` rustdoc::bare_urls warning

Leinnan committed:
Remove Cargo.lock

Leinnan commented:
Responding here: 

> @Leinnan docs.rs still does not work: https://docs.rs/crate/bevy_ios_safearea/0.1.3/builds/1624110

@extrawurst I was looking at it, it is strange, it fails because of the `tracing-oslog` dep. In the new PR I've removed the `Cargo.lock` file from repo and added it to .gitignore. Maybe that would help with that dependency getting into docs build.

extrawurst commented on .gitignore:
```diff
@@ -6,4 +6,4 @@ DerivedData/
 .swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata
 .netrc
-rust/bevy_ios_notifications/Cargo.lock
+/bevy_ios_safearea/Cargo.lock
```
not needed. there is a gitignore right in the subfolder

Leinnan replied:
Done

extrawurst commented on README.md:
````diff
@@ -46,13 +47,20 @@ bevy_ios_safearea = { version = "=0.1.0" }
+
+```rust
+use bevy::prelude::*;
+use bevy_ios_safearea::*;
````
lets `use` only the one type needed to import `IosSafeArea`

Leinnan replied:
It also needs to include the `SafeAreaHelper` AFAIK

extrawurst replied:
lets import that properly or build a `prelude` (the latter is not worth it yet for only two types

Leinnan replied:
Done

extrawurst commented on bevy_ios_safearea/src/lib.rs:
```diff
@@ -1,7 +1,7 @@
-#![doc = include_str!("../README.md")]
+#![doc = include_str!("../../README.md")]
```
this will not work and is working as expected on main

Leinnan replied:
Restored- my solution worked on Windows and main one doesn't. But on the other systems solution from main worked just fine. 
Multiplatform is hard. :(

extrawurst commented on bevy_ios_safearea/src/plugin.rs:
```diff
@@ -1,35 +1,78 @@
-    pub safe_area: Option<IosSafeArea>,
+#[allow(dead_code)]
+/// A trait providing helper methods to access the safe area insets.
+pub trait SafeAreaHelper {
```
* why put this in a trait instead of streight `impl` on `IosSafeAreaResource` (or now `IosSafeArea`)?
* lets name consistently: `IosSafeAreaHelper`

extrawurst replied:
ok now i get it

extrawurst commented on bevy_ios_safearea/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "bevy_ios_safearea"
-version = "0.1.3"
+version = "0.2.0"
```
please update also the changelog

Leinnan replied:
Done

Leinnan committed:
Make docs default target same as in other crates

Leinnan committed:
doc readme path change

Leinnan committed:
Gitignore update

Leinnan committed:
Specify dependencies

Leinnan committed:
Merge branch 'main' into main

Leinnan committed:
Documentation, changelog

Leinnan committed:
Merge branch 'main' of https://github.com/Leinnan/bevy_ios_safearea

extrawurst commented on bevy_ios_safearea/Cargo.toml:
```diff
@@ -10,11 +10,15 @@ keywords = ["bevy", "gamedev", "mobile", "ios", "swift"]
 
 [package.metadata.docs.rs]
-default-target = "x86_64-apple-ios"
+default-target = "x86_64-apple-darwin"
```
this does not make sense to me as our features configs are using ios and not macos. if this is to test if it fixes docs.rs lets do that separately

Leinnan replied:
Let me do a 0.1.5 then xD

extrawurst commented on bevy_ios_safearea/src/plugin.rs:
```diff
@@ -1,35 +1,78 @@
-    pub safe_area: Option<IosSafeArea>,
+#[allow(dead_code)]
+/// A trait providing helper methods to access the safe area insets.
+pub trait SafeAreaHelper {
```
lets name consistently: `IosSafeAreaHelper`

extrawurst commented on CHANGELOG.md:
```diff
@@ -7,6 +7,25 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+### Fixed
+
+* docs.rs fix
```
bold claim :)

extrawurst commented on CHANGELOG.md:
```diff
@@ -7,6 +7,25 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+### Added
+
+* `left` and `right` fields support in `IosSafeArea`
```
best practice is one entry per PR (this also emphasizes splitting into logical PRs)

tigerplush commented on bevy_ios_safearea/src/plugin.rs:
````diff
@@ -1,35 +1,90 @@
+/// }
+// ```
+#[derive(Resource, Clone, Debug, Default, Reflect)]
+#[reflect(Resource)]
````
`register_type::`, `#reflect` and `#derive(Reflect)` all depend on `bevy_reflect`, which isn't enabled in the cargo.toml. I would suggest to have it as optional feature like https://github.com/benfrankel/pyri_tooltip/blob/main/Cargo.toml does

Leinnan replied:
Good call

Leinnan replied:
I will remove it for now, it can be done in seperate PR.

Leinnan committed:
Add Ios prefix, remove reflect stuff from this PR

Leinnan committed:
Merge branch 'main' into main

Leinnan committed:
Docs fix

Leinnan committed:
Merge branch 'main' of https://github.com/Leinnan/bevy_ios_safearea

Leinnan committed:
Merge branch 'main' into main

Leinnan committed:
Test

Leinnan committed:
Cache and format

Leinnan committed:
Merge branch 'main' into main

extrawurst reviewed (changes requested):
ok final review. ideally this is 3 PRs: support left/right, change convenience method to use helper trait and the resurce rename. and then each PR gets one entry in the changelog.
as a bonus you do not make PRs from your main as this prevents me to push to your PR branch and i have to ask for every tiny shit in comments :)

Leinnan commented:
@extrawurst Left/right support yeah- it should be a separate PR. For the rest I feel like it is connected so it makes sense to combine those changes. Main reason was removal of the `IosSafeAreaResource`, trait exists to provide same experience as without it.

extrawurst commented:
> removal of the IosSafeAreaResource

not sure i get that, its not a removal if you promote another one to be the resource after :D - i get the different approach of not storing an `Option` but instead make the whole resource optional though

feel free to make this one then about this

Leinnan committed:
Merge branch 'main' into main

Leinnan committed:
Cleanup

Leinnan commented:
I will create a new one with same content, just not on main so you can edit it as well.
